olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp

```diff
@@ -48,117 +48,141 @@ constexpr std::uint32_t kMaxQuadTreeIndexDepth = 4u;
 
 using namespace olp::client;
 
-SubQuadsRequest PrefetchTilesRepository::EffectiveTileKeys(
-    const std::vector<geo::TileKey>& tile_keys, unsigned int min_level,
-    unsigned int max_level) {
-  SubQuadsRequest ret;
-  for (auto tile_key : tile_keys) {
-    auto child_tiles = EffectiveTileKeys(tile_key, min_level, max_level, true);
-    for (auto tile : child_tiles) {
-      // check if child already exist, if so, use the greater depth
-      auto old_child = ret.find(tile.first);
-      if (old_child != ret.end()) {
-        ret[tile.first] = std::max((*old_child).second, tile.second);
-      } else {
-        ret[tile.first] = tile.second;
-      }
+std::vector<geo::TileKey> PrefetchTilesRepository::GetChilds(
+    const geo::TileKey& tile_key) {
+  if (!tile_key.IsValid()) {
+    return {};
+  }
+  std::vector<geo::TileKey> ret;
+  for (std::uint8_t index = 0; index < kMaxQuadTreeIndexDepth; ++index) {
+    auto tile = tile_key.GetChild(index);
+    if (tile_key.IsValid()) {
+      ret.emplace_back(std::move(tile));
     }
   }
-
   return ret;
 }
 
-SubQuadsRequest PrefetchTilesRepository::EffectiveTileKeys(
-    const geo::TileKey& tile_key, unsigned int min_level,
-    unsigned int max_level, bool add_ancestors) {
-  SubQuadsRequest ret;
-  auto current_level = tile_key.Level();
+std::vector<geo::TileKey> PrefetchTilesRepository::GetChildsAt4LevelsDown(
+    const geo::TileKey& tile_key) {
+  int levels = kMaxQuadTreeIndexDepth + 1;
+  std::vector<geo::TileKey> childrens;
+  std::vector<geo::TileKey> tiles{tile_key};
+  childrens.reserve(std::pow(4, 4));
+  tiles.reserve(std::pow(4, 4));
+
+  while (levels-- > 0) {
+    for (const auto& child : tiles) {
+      auto tile_childrens = GetChilds(child);
+      childrens.insert(childrens.end(),
+                       std::make_move_iterator(tile_childrens.begin()),
+                       std::make_move_iterator(tile_childrens.end()));
+    }
+    if (childrens.empty()) {
+      break;
+    }
+    tiles = childrens;
+    childrens.clear();
+  }
+  return tiles;
+}
```

Comment:
Returned 999 tiles for TileKey(0, 0, 1)
Logic is broken, please add unit tests

Reply:
will be in separate pull request. Logic are changed for slicing from max level to min. It will reduce number of subquads requests.